baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,354 @@
+        printFormat : str, optional
+            Format string valid for use with the str.format() method (e.g "{:17.11e}" for a float or "{:03d}" for an
+            int), by default None, in which case a default format for the given `varType` is used, not used if
+            `printVar` is False
+        """
+        self.variables[name] = {"print": printVar, "type": varType}
+
+        # Create variable data list
+        self.data[name] = []
+
+        # Only store variable's string format if it's going to be printed
+        if printVar:
+            if printFormat is not None:
+                self.variables[name]["format"] = printFormat
+            elif varType in self.defaultFormat:
+                self.variables[name]["format"] = self.defaultFormat[varType]
+            else:
+                self.variables[name]["format"] = self.DEFAULT_OTHER_FORMAT
+
+            # Get test value for figuring out how long a string is using the supplied format
+            try:
+                testValue = self.testValues[varType]
+            except KeyError:
+                testValue = self.DEFAULT_OTHER_VALUE
+
+            # Figure out column width, the maximum of the length of the name and the formatted value, also check that
+            # the format string is valid
+            try:
+                testString = self.variables[name]["format"].format(testValue)
+            except (ValueError, TypeError) as e:
+                raise ValueError(f"Supplied format string \"{self.variables[name]['format']}\" is invalid") from e
+
+            dataLen = len(testString)
+            nameLen = len(name)
+            self.variables[name]["columnWidth"] = max(dataLen, nameLen)
+
+            # --- Now figure out the format strings for the iteration printout header and values ---
+            # The header is simply centred in the available width, which is actually columnWidth + 4
+            self.variables[name]["headerFormat"] = f"{{:^{(self.variables[name]['columnWidth'] + 4)}s}}"
+
+    def startTiming(self) -> None:
+        """Record the start time of the solver
+
+        This function only needs to be called explicitly if the start time of your solver is separate from the first
+        time the `write` method is called.
+        """
+        self.startTime = time.time()
+
+    def write(self, data: dict) -> None:
+        """Record data for a single iteration
+
+        Note that each call to this method is treated as a new iteration. All data to be recorded for an iteration must
+        therefore be recorded in a single call to this method.
+
+        Parameters
+        ----------
+        data : dict
+            Dictionary of values to record, with variable names as keys
+        """
+
+        # Store time
+        if self.startTime < 0.0:
+            self.startTiming()
+            data["Time"] = 0.0
+        else:
+            data["Time"] = time.time() - self.startTime
```

Comment:
Is this time intended to be the solver time, or just some reference time? I think the solver time is something that the solver should manage and provide as data to be written (same as iter). 
Here, there is an additional time added to the actual solver, as this function is potentially called somewhat later after all data has been gathered from the solver to be written here.

Reply:
Yeah not sure, more than the `Iter`, I feel that the solver should probably manage this as there could be some delay until the data is actually written. The delta here can also be different between the calls `write` calls, even though the actual simulation time is the same. Not sure...

Reply:
ok, they are both enabled by default but the user now has the option to completely exclude the `Iter` and `Time` variables from the history if they want. They can always define their own in place of or on top of the default ones.